### ISSUE TITLE
Unable to set "undefined" argument value for headers exchange binding

### ISSUE THREAD

**Issue by shark300** [rabbitmq/discussions#161]
* RabbitMQ version: 3.8.3
* Erlang version: 21.3.8.15
* RabbitMQ server and client application log files:
> 2020-11-06 16:59:26.191 [error] <0.13675.51> Ranch listener rabbit_web_dispatch_sup_15672, connection process <0.13675.51>, stream 1 had its request process <0.13674.51> exit with reason {nocatch,{error,null_not_allowed}} and stacktrace [{rabbit_misc,type_val,1,[{file,"src/rabbit_misc.erl"},{line,417}]},{rabbit_misc,to_amqp_table_row,2,[{file,"src/rabbit_misc.erl"},{line,404}]},{rabbit_misc,'-to_amqp_table/1-fun-0-',3,[{file,"src/rabbit_misc.erl"},{line,398}]},{maps,fold_1,3,[{file,"maps.erl"},{line,257}]},{rabbit_misc,to_amqp_table,1,[{file,"src/rabbit_misc.erl"},{line,398}]},{rabbit_mgmt_format,format_args,1,[{file,"src/rabbit_mgmt_format.erl"},{line,98}]},{rabbit_mgmt_format,'-format/2-lc$^1/1-1-',2,[{file,"src/rabbit_mgmt_format.erl"},{line,52}]},{rabbit_mgmt_util,props_to_method,2,[{file,"src/rabbit_mgmt_util.erl"},{line,909}]}]
> 2020-11-06 16:59:26.190 [error] <0.13674.51> CRASH REPORT Process <0.13674.51> with 0 neighbours crashed with reason: {error,null_not_allowed} in rabbit_misc:type_val/1 line 417
* A runnable code sample, terminal transcript or detailed set of
  instructions that can be used to reproduce the issue

```bash
curl --location --request POST 'https://rabbitmq-test.some.domain/api/bindings/sample-vhost/e/sample.ex/q/sample.q/' \
--header 'Authorization: Basic c2FtcGxlLW1hbjpwYXNzd29yZA==' \
--header 'Content-Type: application/json' \
--data-raw '{"routing_key":"", "arguments":{"type": null}}'`
```
```java
@Bean
public Binding binding(
    HeadersExchange exchange, Queue queue) {
  return BindingBuilder.bind(queue)
      .to(exchange)
      .where("type")
      .exists();
}
```

* RabbitMQ plugin information via `rabbitmq-plugins list`
  * rabbitmq_management
* Client library version (for all libraries used)
  * spring-amqp-2.2.10-RELEASE
  * spring-rabbit-2.2.10-RELEASE
* Operating system, version, and patch level
  * RedHat

When a binding is created by Spring, arguments could be undefined, but it is not possible to configure on UI or by REST.

**Comment by shark300** [comment]
@michaelklishin 

I'm trying to figure out Spring and RabbitMQ internal operation, and as I see, Spring works great. 
If I configure headers exchange with exist() method, it will create (somehow) a working binding and I can't do that with UI.

```java
      .where("type")
      .exists();
```
match:
![exists_by_spring](https://user-images.githubusercontent.com/1255546/98415209-1fb6b300-207d-11eb-8d5c-c49d851889ed.PNG)
no match:
![exists_by_spring_no_matching](https://user-images.githubusercontent.com/1255546/98415210-204f4980-207d-11eb-980f-90fc323d4a2e.PNG)

Unfortunately, I can not configure on UI:

empty string on UI:
![empty_string_by_ui](https://user-images.githubusercontent.com/1255546/98415330-673d3f00-207d-11eb-9dcc-bb88c0c436dd.PNG)
"undefined" string on UI:
![undefined_as_string_by_ui](https://user-images.githubusercontent.com/1255546/98415331-67d5d580-207d-11eb-80e0-34bea8f7230b.PNG)

Last but not least, I configured Spring like this:

```java
        .where("type")
        .matches("undefined")
```
 I've got a same result as I expected.

![undefined_as_string_by_spring](https://user-images.githubusercontent.com/1255546/98415668-fc403800-207d-11eb-82e8-23b607788ff8.PNG)

I've tried to export working configuration by this command: rabbitmqctl export_definitions C:\Temp\definitions.file.json (I have configured a local server RabbitMQ 3.8.9, Erlang 23.1.2) but there is no argument have been exported (or imported as well):

[definitions.file.json.zip](https://github.com/rabbitmq/discussions/files/5503567/definitions.file.json.zip)

So to recap, I really don't know how, but Spring can configure "existing" operation on headers exchange, and it works. Management UI can't reproduce that, and I can't export this configurations too.

**Comment by shark300** [comment]
I've created a quick sample project too.

https://github.com/shark300/rabbitmq_161

**Comment by michaelklishin** [comment]
A [traffic capture](https://www.rabbitmq.com/amqp-wireshark.html) would tell you exactly what your Spring code does. Are you sure that it uses the HTTP API to set up the bindings, for example? In any case, matching on `null` will not achieve what you likely want. Headers exchanges cannot match on absence of a key.

**Comment by michaelklishin** [comment]
If there is no argument in the definitions, it means that it's skipped by this Spring code. That would explain why it does not fail validation. The solution to replicate this via the HTTP API would be the same: skip the null value. I have demonstrated above using `curl` that it works as expected.

**Comment by shark300** [comment]
@michaelklishin 
Good catch, Spring is using AMQP to declare exchanges/queues/bindings; even so, it should be configurable on UI without Spring Admin.
Thank you for your tips. I've captured the traffic by Wireshark.

![image](https://user-images.githubusercontent.com/1255546/98451879-62928c80-214a-11eb-924b-6c37c353a5a5.png)

[rabbitmq_capture.zip](https://github.com/rabbitmq/discussions/files/5505470/rabbitmq_capture.zip)

**Comment by michaelklishin** [comment]
What *exactly* should be configurable in the management UI? There is a way to configure arguments but `null` values are explicitly unsupported and I explained why. Any Web UI form, at least in the current interface, has to worry about blank fields that were never touched by the user, especially because of the [declaration property equivalence requirement](https://www.rabbitmq.com/queues.html#property-equivalence).

I am not convinced that there is any reason for us to remove the validation; we won't do it because of a single reported case: much more demand for this change should emerge first.

**Comment by shark300** [comment]
@michaelklishin 
I see your point, but unfortunately, I can't configure RabbitMQ exchanges on our staging environments via AMQP (for security reason). I have to open a support ticket for configuring our RabbitMQ bindings but our support teams can't do that on **UI** or via **import_definitions**. Due to this there is no way for us to configure headers exchange  argument "exists".

**Comment by shark300** [comment]
Hi @michaelklishin 
Do you have any news about this issue?
If not, we will develop a migration tool for our support team. This tool [can use native AMQP](https://github.com/rabbitmq/discussions/issues/161#issuecomment-723498984) to modify headers's binding instead of calling REST API or importing JSON.
